### PR TITLE
Fix: Logger attachments not coming in vsts test run

### PR DESCRIPTION
**Pull request by abhishkk**
## Description
In vsts, HandleRawMessage was sending complete event before HandleTestRunComplete can call loggers. Thus attachments created by loggers were not coming

## Fix
Similar to data collector, we have moved loggers invokation to HandleRawMessage.

### CODE REVIEW

File: src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs [hunk 1]
```diff
@@ -394,54 +393,94 @@ public void HandleLogMessage(TestMessageLevel level, string message)
...
+
+            var message = this.LoggerManager.LoggersInitialized || this.requestData.IsTelemetryOptedIn ?
+                    this.dataSerializer.DeserializeMessage(rawMessage) : null;
+
+            if (string.Equals(message?.MessageType, MessageType.DiscoveryComplete))
             {
-                var message = this.dataSerializer.DeserializeMessage(rawMessage);
+                var discoveryCompletePayload = this.LoggerManager.LoggersInitialized || this.requestData.IsTelemetryOptedIn ?
+                    this.dataSerializer.DeserializePayload<DiscoveryCompletePayload>(message) : default(DiscoveryCompletePayload);
+                rawMessage = UpdateRawMessageWithTelemetryInfo(discoveryCompletePayload, message) ?? rawMessage;
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs]: >UpdateRawMessageWithTelemetryInfo [](start = 29, length = 33)

what if telemetry is disabled, do we still want to add telemetry data to message?

File: src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs [hunk 2]
```diff
@@ -394,54 +393,94 @@ public void HandleLogMessage(TestMessageLevel level, string message)
...
-            if (this.requestData.IsTelemetryOptedIn)
+            // Note: Deserialize rawMessage only if required.
+
+            var message = this.LoggerManager.LoggersInitialized || this.requestData.IsTelemetryOptedIn ?
+                    this.dataSerializer.DeserializeMessage(rawMessage) : null;
+
+            if (string.Equals(message?.MessageType, MessageType.DiscoveryComplete))
             {
-                var message = this.dataSerializer.DeserializeMessage(rawMessage);
+                var discoveryCompletePayload = this.LoggerManager.LoggersInitialized || this.requestData.IsTelemetryOptedIn ?
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs]: >this.LoggerManager.LoggersInitialized || this.requestData.IsTel [](start = 47, length = 63)

are these checks needed again? You have already checked above that either Loggers or telemetry is enabled

File: src/Microsoft.TestPlatform.Client/Execution/TestRunRequest.cs [hunk 3]
```diff
@@ -543,54 +542,98 @@ public void HandleLogMessage(TestMessageLevel level, string message)
...
-                        else
-                        {
-                            foreach (var kvp in this.requestData.MetricsCollection.Metrics)
-                            {
-                                testRunCompletePayload.TestRunCompleteArgs.Metrics[kvp.Key] = kvp.Value;
-                            }
-                        }
+            if (string.Equals(message?.MessageType, MessageType.ExecutionComplete))
+            {
+                var testRunCompletePayload = this.LoggerManager.LoggersInitialized || this.requestData.IsTelemetryOptedIn ?
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Client/Execution/TestRunRequest.cs]: same comment as in DiscoveryRequest

File: src/vstest.console/TestPlatformHelpers/TestRequestManager.cs [hunk 4]
```diff
@@ -427,11 +427,16 @@ private bool UpdateRunSettingsIfRequired(string runsettingsXml, List<string> sou
...
 
+                    // Add or update console logger.
                     if (!designMode)
                     {
                         AddOrUpdateConsoleLogger(document, runsettingsXml);
                         settingsUpdated = true;
                     }
+                    else
+                    {
+                        settingsUpdated = settingsUpdated || UpdateConsoleLoggerIfExists(document, runsettingsXml);
```
Comment (mayankbansal018) [inline src/vstest.console/TestPlatformHelpers/TestRequestManager.cs]: Why do we want to enable Console logger in DesignMode?

File: src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs (hunk 1 above)
Comment (abhishkk) [reply src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs]: No we are not adding telemetry data when its disabled

File: src/vstest.console/TestPlatformHelpers/TestRequestManager.cs (hunk 4 above)
Comment (abhishkk) [reply src/vstest.console/TestPlatformHelpers/TestRequestManager.cs]: If user has added in runsettings, we are allowing console logger.

File: src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs [hunk 5]
```diff
@@ -270,13 +270,12 @@ public void HandleDiscoveryComplete(DiscoveryCompleteEventArgs discoveryComplete
...
                     {
                         var discoveredTestsEvent = new DiscoveredTestsEventArgs(lastChunk);
-                        this.OnDiscoveredTests.SafeInvoke(this, discoveredTestsEvent, "DiscoveryRequest.DiscoveryComplete");
                         this.LoggerManager.HandleDiscoveredTests(discoveredTestsEvent);
+                        this.OnDiscoveredTests.SafeInvoke(this, discoveredTestsEvent, "DiscoveryRequest.DiscoveryComplete");
                     }
 
-                    this.OnDiscoveryComplete.SafeInvoke(this, discoveryCompleteEventArgs, "DiscoveryRequest.DiscoveryComplete");
                     this.LoggerManager.HandleDiscoveryComplete(discoveryCompleteEventArgs);
-                    this.LoggerManager.Dispose();
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs]: Where are we disposing now ?

Comment (abhishkk) [reply src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs]: In LoggerManager.HandleDiscoveryCompelte

File: src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs (hunk 2 above)
Comment (abhishkk) [reply src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs]: removed

File: src/Microsoft.TestPlatform.Client/Execution/TestRunRequest.cs (hunk 3 above)
Comment (abhishkk) [reply src/Microsoft.TestPlatform.Client/Execution/TestRunRequest.cs]: removed

File: src/vstest.console/TestPlatformHelpers/TestRequestManager.cs [hunk 6]
```diff
@@ -440,41 +445,70 @@ private bool UpdateRunSettingsIfRequired(string runsettingsXml, List<string> sou
...
+
+        /// <summary>
+        /// Add console logger in runsettings if exists.
+        /// </summary>
+        /// <param name="document">Runsettings document.</param>
+        /// <param name="runsettingsXml">Runsettings xml.</param>
+        /// <returns>True if updated console logger in runsettings successfully.</returns>
+        private bool UpdateConsoleLoggerIfExists(XmlDocument document, string runsettingsXml)
+        {
             var dummyConsoleLogger = new LoggerSettings
```
Comment (singhsarab) [inline src/vstest.console/TestPlatformHelpers/TestRequestManager.cs]: >dummyConsoleLogger [](start = 16, length = 18)

nit: default*

File: src/Microsoft.TestPlatform.ObjectModel/Client/Events/TestRunCompleteEventArgs.cs [hunk 7]
```diff
@@ -31,7 +31,7 @@ public TestRunCompleteEventArgs(ITestRunStatistics stats, bool isCanceled, bool
             this.IsCanceled = isCanceled;
             this.IsAborted = isAborted;
             this.Error = error;
-            this.AttachmentSets = attachmentSets;
+            this.AttachmentSets = attachmentSets ?? new Collection<AttachmentSet>(); // Ensuring attachmentSets are not null, so that new attachmentSets can be combined whenever required.
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.ObjectModel/Client/Events/TestRunCompleteEventArgs.cs]: Please put a comment why we are now initialization it

File: test/Microsoft.TestPlatform.Client.UnitTests/Discovery/DiscoveryRequestTests.cs [hunk 8]
```diff
@@ -278,5 +222,70 @@ public void HandleRawMessageShouldAddVSTestDataPointsIfTelemetryOptedIn()
...
+
+        [TestMethod]
+        public void HandleDiscoveredTestsShouldInvokeHandleDiscoveredTestsOfLoggerManager()
+        {
+            discoveryRequest.HandleDiscoveredTests(null);
+
+            loggerManager.Verify(lm => lm.HandleDiscoveredTests(It.IsAny<DiscoveredTestsEventArgs>()), Times.Once);
+        }
     }
 }
```
Comment (mayankbansal018) [inline test/Microsoft.TestPlatform.Client.UnitTests/Discovery/DiscoveryRequestTests.cs]: Need test to validate that HandleDiscoveryComplete of Logger Manager is invoked only once,if the message comes from both HandleRawMessage, & HandleDiscoveryComplete

File: test/Microsoft.TestPlatform.Client.UnitTests/Execution/TestRunRequestTests.cs [hunk 9]
```diff
@@ -595,13 +676,5 @@ public void ExecuteAsyncShouldInvokeHandleTestRunStartOfLoggerManager()
...
         }
-
-        [TestMethod]
-        public void HandleLogMessageShouldInvokeHandleLogMessageOfLoggerManager()
-        {
-            testRunRequest.HandleLogMessage(TestMessageLevel.Error, "hello");
-            loggerManager.Verify(lm => lm.HandleTestRunMessage(It.IsAny<TestRunMessageEventArgs>()), Times.Once);
-
-        }
     }
```
Comment (mayankbansal018) [inline test/Microsoft.TestPlatform.Client.UnitTests/Execution/TestRunRequestTests.cs]: Same as in discovery tests

**Review by mayankbansal018 — APPROVED** [review]

File: test/Microsoft.TestPlatform.Client.UnitTests/Discovery/DiscoveryRequestTests.cs (hunk 8 above)
Comment (abhishkk) [reply test/Microsoft.TestPlatform.Client.UnitTests/Discovery/DiscoveryRequestTests.cs]: UT are there for this scenario.

File: test/Microsoft.TestPlatform.Client.UnitTests/Execution/TestRunRequestTests.cs (hunk 9 above)
Comment (abhishkk) [reply test/Microsoft.TestPlatform.Client.UnitTests/Execution/TestRunRequestTests.cs]: UTs are present for this scenario.

**Review by singhsarab — APPROVED** [review]
:shipit:

**Pull request merged**